File: Two_Pointer/Longest_Substring_Without_Repeating_Characters.js
var lengthOfLongestSubstring = function(s) {
    if(s.length <= 1) return s.length;
    
    let res = 1;
    let start = 0, end = 0;
    for(let i=1; i<s.length; i++) {
        if(s[i] !== s[end]) {
            end++;
            let found = false;
            for(let j=start; j<end; j++) {
                if(s[j] === s[i]) {
                    start = j+1;
                    found = true;
                    break;
                }
            }
            if(!found) {
                res = Math.max(end-start+1, res);
            }
        } else {
            start = i; 
            end = i;
        }
    }
    return res;
};

File: Two_Pointer/Longest_Substring_Without_Repeating_Characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        result = 0
        left = 0
        charSet = set()

        for char in s:
            while char in charSet:
                charSet.remove(s[left])
                left += 1

            charSet.add(char)
            result = max(result, len(charSet))

        return result